### PR TITLE
Update/unit test additions

### PR DESCRIPTION
**Pull request by JWZeta**
Adding unit tests for functions:
pack_digest()
unpack_digest()
unseal_apply_policy()
apply_policy_or()
convert_string_to_digest()
convert_digest_to_string()
parse_pcrs_string()

### CODE REVIEW

File: test/src/tpm/formatting_tools_test.c [hunk 1]
```diff
@@ -2136,3 +2149,59 @@ void test_concat(void)
...
+  string = (char *) malloc((digest.size *2) + 1);
+
+  CU_ASSERT(convert_digest_to_string(&digest, string) == 0);
+  CU_ASSERT(string != NULL);
+  CU_ASSERT(convert_string_to_digest(string, &test_digest) == 0);
+  CU_ASSERT(test_digest.size != 0);
+  free(string);
+}
+
+//----------------------------------------------------------------------------
```
Comment (bit-bender) [inline test/src/tpm/formatting_tools_test.c:2178]: This is really just an observation, but for some reason, we put the marshal/unmarshal (pack/unpack) functionality in a separate source file (marshalling_tools.c), but, we combined these tests with those for formatting_tools.c in formatting_tools_test.c instead of creating a marshalling_tools_test.c (this includes tests for utilities that are not part of this pull request).

File: test/src/tpm/tpm2_interface_test.c [hunk 2]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
   CU_ASSERT(rollNonces(&session, new) != 0);
 
 }
+
+
+
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:766]:
```suggestion
```

File: test/src/tpm/tpm2_interface_test.c [hunk 3]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
...
+// test_unseal_apply_policy
+//----------------------------------------------------------------------------
+void test_unseal_apply_policy(void)
+{
+  TSS2_SYS_CONTEXT *sapi_ctx = NULL;
+
+  init_tpm2_connection(&sapi_ctx);
+  TPML_PCR_SELECTION pcrs_struct = {.count = 0, };
+
+
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:777]:
```suggestion
```

File: test/src/tpm/tpm2_interface_test.c (hunk 2 above)
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:766]:
```suggestion
```

File: test/src/tpm/tpm2_interface_test.c [hunk 4]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
   CU_ASSERT(rollNonces(&session, new) != 0);
 
 }
+
+
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:765]:
```suggestion
```

File: test/src/tpm/tpm2_interface_test.c (hunk 2 above)
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:766]:
```suggestion
```

File: test/src/tpm/tpm2_interface_test.c [hunk 5]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
...
+
+//----------------------------------------------------------------------------
+// test_unseal_apply_policy
+//----------------------------------------------------------------------------
+void test_unseal_apply_policy(void)
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:770]: This test assumes that the default hash algorithm is SHA-256. If this is ever changed, I believe that these tests will break. Perhaps we should put a conditional around the current tests to only run them if the expected default value is set and warn that tests were skipped if the default is something we didn't expect.

```c
if (KMYTH_HASH_ALG == TPM2_ALG_SHA256)
{
    [tests]
}
else
{
    [warning]
}
```

File: test/src/tpm/tpm2_interface_test.c [hunk 6]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
...
+  policy2.size = 0;
+  policyOR.size = 0;
+
+  int pcrs[1] = { };
+  pcrs[0] = 23;
+  init_pcr_selection(sapi_ctx, pcrs, 1, &pcrs_struct);
+  CU_ASSERT(create_policy_digest(sapi_ctx, pcrs_struct, &policy1) == 0);
+  CU_ASSERT(policy1.size != 0);
+  //Remove printf and tpm2_pcrread after review
+  printf("\nPolicy1: %X\n", policy1.buffer);
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:839]: This in not part of what will be merged and doesn't need to be changed, but the 8-byte value printed here is not the 32-byte digest. There may be a more elegant way to do it, but I think something like this will work:

```  printf("\nPolicy1: 0x");
  for (int i = 0; i < policy1.size; i++)
  {
    printf("%02X", policy1.buffer[i]);
  }
  printf("\n");
```

File: test/src/tpm/tpm2_interface_test.c [hunk 7]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
...
+  system("tpm2_pcrread sha256:23");
+
+  system("tpm2_pcrextend 23:sha256=0000000000000000000000000000000000000000000000000000000000000001");
+  init_pcr_selection(sapi_ctx, pcrs, 1, &pcrs_struct);
+  CU_ASSERT(create_policy_digest(sapi_ctx, pcrs_struct, &policy2) == 0);
+  CU_ASSERT(policy2.size != 0);
+  //Remove printf and tpm2_pcrread after review
+  printf("Policy2: %X\n", policy2.buffer);
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:847]: Same observation-only comment that this does not print the policy digest value.

File: test/src/tpm/tpm2_interface_test.c [hunk 8]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
...
+  TPML_DIGEST pHashList;
+  SESSION policySessionOR;
+  create_auth_session(sapi_ctx, &policySessionOR, TPM2_SE_TRIAL);
+  CU_ASSERT(apply_policy_or(sapi_ctx, policySessionOR.sessionHandle, &policy1,
+                    &policy2, &pHashList) == 0);
+  CU_ASSERT(Tss2_Sys_PolicyGetDigest(sapi_ctx, policySessionOR.sessionHandle,
+                             nullCmdAuths, &policyOR, nullRspAuths) == 0);
+  CU_ASSERT(policyOR.size != 0);
+  //Remove printf after review
+  printf("PolicyOR: %X\n", policyOR.buffer);
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:859]: Same observation-only comment that this does not print the policy digest value.

File: test/src/tpm/tpm2_interface_test.c [hunk 9]
```diff
@@ -749,3 +761,104 @@ void test_rollNonces(void)
...
+  int pcrs[1] = { };
+  pcrs[0] = 23;
+  init_pcr_selection(sapi_ctx, pcrs, 1, &pcrs_struct);
+  CU_ASSERT(create_policy_digest(sapi_ctx, pcrs_struct, &policy1) == 0);
+  CU_ASSERT(policy1.size != 0);
+  //Remove printf and tpm2_pcrread after review
+  printf("\nPolicy1: %X\n", policy1.buffer);
+  system("tpm2_pcrread sha256:23");
+
+  system("tpm2_pcrextend 23:sha256=0000000000000000000000000000000000000000000000000000000000000001");
```
Comment (bit-bender) [inline test/src/tpm/tpm2_interface_test.c:842]: If this TPM2 Tools command fails, subsequent tests should fail because of improper test setup. Perhaps, though, we could instead check the return value from the system call and if it fails, return a more targeted test fail result:

```c
  if (system("tpm2_pcrextend 23:sha256=000000000000000000000000000000000000000000000000000000000000001") == 0)
  {
    init_pcr_selection(sapi_ctx, pcrs, 1, &pcrs_struct);
    CU_ASSERT(create_policy_digest(sapi_ctx, pcrs_struct, &policy2) == 0);
    CU_ASSERT(policy2.size != 0);
    //Remove printf and tpm2_pcrread after review
    printf("\nPolicy2: 0x");
    for (int i = 0; i < policy2.size; i++)
    {
      printf("%02X", policy2.buffer[i]);
    }
    printf("\n");
    printf("Policy2: %X\n", policy2.buffer);
    system("tpm2_pcrread sha256:23");
  
    TPML_DIGEST pHashList;
    SESSION policySessionOR;
    create_auth_session(sapi_ctx, &policySessionOR, TPM2_SE_TRIAL);
    CU_ASSERT(apply_policy_or(sapi_ctx, policySessionOR.sessionHandle, &policy1,
                      &policy2, &pHashList) == 0);
    CU_ASSERT(Tss2_Sys_PolicyGetDigest(sapi_ctx, policySessionOR.sessionHandle,
                               nullCmdAuths, &policyOR, nullRspAuths) == 0);
    CU_ASSERT(policyOR.size != 0);
    //Remove printf after review
    printf("\nPolicyOR: 0x");
    for (int i = 0; i < policyOR.size; i++)
    {
      printf("%02X", policyOR.buffer[i]);
    }
    printf("\n");
    printf("PolicyOR: %X\n", policyOR.buffer);
    Tss2_Sys_FlushContext(sapi_ctx, policySessionOR.sessionHandle);
  
    system("tpm2_pcrreset 23");
  }
  else
  {
    CU_FAIL("TPM2 Tools (tpm2_pcrextend) system call setting up test failed");
  }
```

**Review by bit-bender — COMMENTED** [review]
Please see comments.

File: test/include/tpm/formatting_tools_test.h [hunk 10]
```diff
@@ -112,5 +112,7 @@ void test_create_nkl_bytes(void);
 void test_encodeBase64Data(void);
 void test_decodeBase64Data(void);
 void test_concat(void);
+void test_verifyConvertionStringDigest(void);
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply test/include/tpm/formatting_tools_test.h:115]: Changed the function to 'test_verifyStringDigestConversion'.

File: test/src/tpm/formatting_tools_test.c [hunk 11]
```diff
@@ -2136,3 +2149,59 @@ void test_concat(void)
   CU_ASSERT(concat(&dest, &dest_len, chile, -1) == 1);
   free(dest);
 }
+
+//----------------------------------------------------------------------------
+// test_verifyConvertionStringDigest()
+//----------------------------------------------------------------------------
+void test_verifyConvertionStringDigest(void)
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply test/src/tpm/formatting_tools_test.c]: Added the stronger test and left the printf() statements for now.

File: test/src/tpm/tpm2_interface_test.c (hunk 9 above)
Comment (JWZeta) [reply test/src/tpm/tpm2_interface_test.c]: Added the system check and fail statement.

File: test/src/tpm/tpm2_interface_test.c (hunk 5 above)
Comment (JWZeta) [reply test/src/tpm/tpm2_interface_test.c:768]: Added check

File: test/src/tpm/formatting_tools_test.c [hunk 12]
```diff
@@ -2136,3 +2149,59 @@ void test_concat(void)
...
+
+  uint8_t * packed_data;
+  size_t packed_data_size = 0;
+  size_t packed_data_offset = 0;
+
+  create_policy_digest(sapi_ctx, pcrs_struct, &digest);
+  CU_ASSERT(digest.size != 0);
+
+  packed_data_size = (digest.size * 2) + 1;
+  packed_data = (uint8_t *) malloc(packed_data_size);
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply test/src/tpm/formatting_tools_test.c:2227]: Added free

File: test/src/tpm/formatting_tools_test.c [hunk 13]
```diff
@@ -2136,3 +2149,59 @@ void test_concat(void)
...
+
+  CU_ASSERT(pack_digest(&digest, packed_data, packed_data_size, packed_data_offset) == 0);
+  CU_ASSERT(packed_data != NULL);
+  CU_ASSERT(unpack_digest(&digest_out, packed_data, packed_data_size, packed_data_offset) == 0);
+  CU_ASSERT(digest_out.size != 0);
+  CU_ASSERT(digest_out.size = digest.size);
```
[the remark this answers is not in the document]
Comment (JWZeta) [reply test/src/tpm/formatting_tools_test.c:2233]: Added verification of results matches

File: test/src/tpm/formatting_tools_test.c (hunk 1 above)
Comment (JWZeta) [reply test/src/tpm/formatting_tools_test.c:2206]: The marshaling _tools.c used to be part of the formatting_tools.c. We should update the tests to ad a marshalling_tools_test.c. I have submitted a issue to do this.
